csrc/scheduler/vectorize_helper.cpp

```diff
@@ -229,148 +229,192 @@ void ContiguousInnerDimensionsMapper::distributePE(
+            pos_after_outer != inner_pos) {
+          // Skip broadcast axes as they must not have been concretized in
+          // the reference. We remove dimensions that underwent a
+          // concretization as well as the dimensions to the left of that in
+          // propagateC2P.
+          return;
         }
+        break;
       }
 
-      ids[inner_pos] = split->in();
-      bool outer_mapped = find_outer_it != ids.end();
-      if (outer_mapped) {
-        // Remove outer
-        ids.erase(find_outer_it);
-      } else {
-        // Clear to the left of inner in since only inner maps
-        ids.erase(ids.begin(), ids.begin() + (int64_t)inner_pos);
+      if (inner_pos != pos_after_outer) {
+        // Nothing to the left of inner could be continuous.
+        //
+        // Clear left of inner
+        clear_left_of(find_inner_it);
       }
+    }
 
-      combinePE(split, outer_mapped);
-    } else if (const Merge* merge = dynamic_cast<const Merge*>(expr)) {
-      auto find_out_it = std::find(ids.begin(), ids.end(), merge->out());
-      if (find_out_it == ids.end()) {
-        continue;
-      }
+    if constexpr (std::is_same_v<decltype(merge_or_split), Split*>) {
+      frontier[inner_pos] = merge_or_split->in();
+    } else {
+      static_assert(std::is_same_v<decltype(merge_or_split), Merge*>);
+      frontier[inner_pos] = merge_or_split->out();
+    }
+    bool outer_mapped = find_outer_it != frontier.end();
+    if (outer_mapped) {
+      // Remove outer
+      frontier.erase(find_outer_it);
+    } else {
+      // Clear to the left of inner in since only inner maps
+      frontier.erase(frontier.begin(), frontier.begin() + (int64_t)inner_pos);
+    }
+
+    combinePE(merge_or_split, outer_mapped);
+  };
+
+  // Process `merge_or_split` and update `frontier`, where `merge_or_split` must
+  // be an expr that unflatten one ID into two IDs. For forward propagation, it
+  // must be a split, and for backward propagation, it must be a merge
+  auto propagateDistribute = [&frontier, this](auto* merge_or_split) {
+    Val* combined = nullptr;
+    if constexpr (std::is_same_v<decltype(merge_or_split), Split*>) {
+      combined = merge_or_split->in();
+    } else {
+      static_assert(std::is_same_v<decltype(merge_or_split), Merge*>);
+      combined = merge_or_split->out();
+    }
+    auto find_out_it = std::find(frontier.begin(), frontier.end(), combined);
+    if (find_out_it == frontier.end()) {
+      return;
+    }
+
+    auto out_pos = std::distance(frontier.begin(), find_out_it);
+    frontier[out_pos] = merge_or_split->outer();
+    frontier.insert(frontier.begin() + out_pos + 1, merge_or_split->inner());
+
+    distributePE(merge_or_split);
+  };
+
+  auto clear_left_of = [&frontier](IterDomain* id) {
+    auto it = std::find(frontier.begin(), frontier.end(), id);
+    if (it != frontier.end()) {
+      frontier.erase(frontier.begin(), it + 1);
+    }
+  };
+
+  auto backward_exprs = StmtSort::getExprsBetween(
```

Comment:
Using `getExprsBetween` from the `to` vector to the `frontier` vector is to handle the case where the `from` parameter IDs are (direct or indirect) consumers of the `to` parameter IDs. If not, this just returns an empty vector, so nothing is done here. For the case, the other traversal below takes care of it. Is this correct?

Reply:
Yes, exactly. Similar things is done in:
https://github.com/NVIDIA/Fuser/blob/329aafed2d6d2998d5e1372ebf08633ee96c0828/csrc/executor_kernel_arg.cpp#L250-L251

Reply:
Thanks. Can you please leave it as a comment?

Reply:
added